### PR TITLE
Changes to support TNLRV3 fine-tuning

### PR DESCRIPTION
**Pull request by Tixxx**
**Description**: Describe your changes.
Changes to support tnlrv3 fine-tuning task
1. added gradient op for ReduceLogSumExp, reference implementation taken from pytorch [here](https://github.com/pytorch/pytorch/blob/0a0960126cd87a6cef04f0ddf9154854b7bd03a0/tools/autograd/templates/Functions.cpp#L493) 
2. fixed bug when passing fp16 to Cudnnreduce kernel. runtime type should be float
3. added sanitization code in python frontend to remove redundant states to match with pytorch state dict.

**Motivation and Context**
To support tnlrv3 fine-tuning

### CODE REVIEW

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 1]
```diff
@@ -788,6 +788,35 @@ IMPLEMENT_GRADIENT_BUILDER(GetReduceMeanGradient) {
...
+  if (attributes.find("keepdims") != attributes.end() &&
+      attributes.at("keepdims").has_i()) {
+    keepdims = static_cast<bool>(attributes.at("keepdims").i());
+  }
+
+  ArgDef grad = GO(0);
+  if (!keepdims && attributes.find("axes") != attributes.end()) {
+    std::vector<int64_t> axes_values = RetrieveValues<int64_t>(attributes.at("axes"));
+    grad = IA("Unsqueezed_Grad");
+    result.push_back(NodeDef("Unsqueeze", {GO(0)}, {grad}, {MakeAttribute("axes", axes_values)}));
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder.cc:804]: I think neither Unsqueeze are needed here, since Sub and Mul both supports broadcasting.

File: orttraining/orttraining/python/ort_trainer.py [hunk 2]
```diff
@@ -773,7 +773,13 @@ def state_dict(self):
             if n.name not in torch_state:
                 torch_state[n.name] = torch.from_numpy(numpy_helper.to_array(n))
 
-        return torch_state
+        # Need to remove redundant initializers and name suffices to map back to original torch state names
+        torch_state_to_return = {}
+        for name, value in torch_state.items():
+            if not (("Moment" in name) or ("Update_Count" in name)):
+                name = name.replace('_fp16', '')
```
Comment (SherlockNoMad) [inline orttraining/orttraining/python/ort_trainer.py:780]: Are we saving fp16 version of the model weights in pytorch state_dict? 
I am wondering why it's needed?

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 3]
```diff
@@ -788,6 +788,35 @@ IMPLEMENT_GRADIENT_BUILDER(GetReduceMeanGradient) {
...
+
+    result.push_back(NodeDef("Unsqueeze", {O(0)}, {IA("Unsqueezed_Output")}, {MakeAttribute("axes", axes_values)}));
+    result.push_back(NodeDef("Sub", {I(0), IA("Unsqueezed_Output")}, {IA("Self_Sub_Result")}));
+  }
+  else {
+    result.push_back(NodeDef("Sub", {I(0), O(0)}, {IA("Self_Sub_Result")}));
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder.cc:810]: Please mark the math here

'''dx_i = exp(xi) / sum(exp(xi)) 

O(0) = log(sum(exp(xi)))
I(0) - O(0) = xi - log(sum(exp(xi))) = log( xi / sum(exp(xi)) )
'''

File: orttraining/orttraining/test/gradient/gradient_ops_test.cc [hunk 4]
```diff
@@ -546,6 +546,74 @@ TEST(GradientCheckerTest, ReduceSumGrad) {
   }
 }
 
+TEST(GradientCheckerTest, ReduceLogSumExpGrad) {
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/gradient/gradient_ops_test.cc:549]: Follow up: I think we can refactor the test code and make a shared function for ReduceSum, ReduceMean and ReduceLogSumExp

File: orttraining/orttraining/python/ort_trainer.py (hunk 2 above)
Comment (SherlockNoMad) [reply orttraining/orttraining/python/ort_trainer.py:780]: BTW, not a big fan of implicit name matching. 
Can you explore if we can we this explicit mapping?

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 5]
```diff
@@ -788,6 +788,35 @@ IMPLEMENT_GRADIENT_BUILDER(GetReduceMeanGradient) {
...
+  }
+
+  result.push_back(NodeDef("Exp", {IA("Self_Sub_Result")}, {IA("Self_Sub_Result_Exp")}));
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder.cc:813]: Add a comment here: potential optimization with kernel fusion, and recompute with only using I(0)

Comment (SherlockNoMad) [reply orttraining/orttraining/core/graph/gradient_builder.cc:813]: This gradient uses both I(0) and O(0) which might cause more stash activation memory

File: onnxruntime/core/providers/cuda/reduction/reduction_ops.cc [hunk 6]
```diff
@@ -396,8 +396,9 @@ Status ReduceComputeCore(CUDAExecutionProvider& cuda_ep, const Tensor& input, Pr
   }
 
   CudnnReduceDescriptor reduce_desc;
-  if (std::is_same<T, MLFloat16>::value)
+  if (std::is_same<T, MLFloat16>::value) {
     ORT_RETURN_IF_ERROR(reduce_desc.Set(cudnn_reduce_op, CudnnTensor::GetDataType<float>(), ReduceTensorIndices));
+  }
   else
```
Comment (SherlockNoMad) [inline onnxruntime/core/providers/cuda/reduction/reduction_ops.cc:402]: } else

File: onnxruntime/core/providers/cuda/reduction/reduction_ops.cc [hunk 7]
```diff
@@ -438,7 +439,11 @@ Status ReduceComputeCore(CUDAExecutionProvider& cuda_ep, const Tensor& input, Pr
       } else {
         // Reduce max -- Max/Min will output indices data
         CudnnReduceDescriptor reduce_max_desc;
-        ORT_RETURN_IF_ERROR(reduce_max_desc.Set(CUDNN_REDUCE_TENSOR_MAX, cudnn_type_X, CUDNN_REDUCE_TENSOR_NO_INDICES));
+        cudnnDataType_t cudnn_reduce_max_type = cudnn_type_X;
+        if((std::is_same<T, MLFloat16>::value)) {
+            cudnn_reduce_max_type = CUDNN_DATA_FLOAT;
```
Comment (SherlockNoMad) [inline onnxruntime/core/providers/cuda/reduction/reduction_ops.cc:444]: This is fixing for ReduceMax? 
I thought we have problem for ReduceLogSumExp? 

Also, I am not sure how well cudnn reduction works for half type, Weixing had some concerned with this before. 
Another approach is to cast is to fp32 first and use cudnn reduction in float.

File: onnxruntime/core/providers/cuda/reduction/reduction_ops.cc (hunk 6 above)
Comment (SherlockNoMad) [reply onnxruntime/core/providers/cuda/reduction/reduction_ops.cc:402]: also fix {} for the else case ?

File: onnxruntime/core/providers/cuda/reduction/reduction_ops.cc (hunk 7 above)
Comment (Tixxx) [reply onnxruntime/core/providers/cuda/reduction/reduction_ops.cc:444]: Yea this whole function is written in a strange way. But the flow is: 1. get the index space for reducing(this is for keepdim since whether to keep dim or not will change output shape) 2. reduceMax on the input and copy result. 3. perform log->reduceSum->exp... Step 1 and step 2 need to be using the same type, in this case, step 1 was using a different type as step 2, it was complaining about type mismatch.. But the runtime type of the actual computation in step 3 is all done in float.

File: orttraining/orttraining/core/graph/gradient_builder.cc (hunk 5 above)
Comment (Tixxx) [reply orttraining/orttraining/core/graph/gradient_builder.cc:813]: Sure. I don't know how intelligent we want the recompute scheduler to be. If we run a dataflow analysis on the graph, a gradient node that uses output from its forward op can most likely qualify as a recompute candidate so that we don't need to annotate node by node.

File: orttraining/orttraining/core/graph/gradient_builder.cc (hunk 1 above)
Comment (Tixxx) [reply orttraining/orttraining/core/graph/gradient_builder.cc:804]: I thought about it, but since this op takes a dims as input essentially users can choose which dim to reduce, we need to reply on which axes to unsqueeze based on the input dims? I just wanted to avoid corner cases since bcast only aligns either the leftmost or the rightmost.

File: orttraining/orttraining/core/graph/gradient_builder.cc (hunk 3 above)
Comment (Tixxx) [reply orttraining/orttraining/core/graph/gradient_builder.cc:810]: sounds good.

File: orttraining/orttraining/core/graph/gradient_builder.cc (hunk 1 above)
Comment (SherlockNoMad) [reply orttraining/orttraining/core/graph/gradient_builder.cc:804]: yes, you are right. 
think of the case [A, B, C] reduced to [A, C] with axes = [1] and keepdim=0, there are actuality not broadcast-able.

File: onnxruntime/core/providers/cuda/reduction/reduction_ops.cc (hunk 7 above)
Comment (SherlockNoMad) [reply onnxruntime/core/providers/cuda/reduction/reduction_ops.cc:444]: I see. ReduceMax is used to avoid numerical problem of doing exp direction on x.

File: orttraining/orttraining/python/ort_trainer.py (hunk 2 above)
Comment (Tixxx) [reply orttraining/orttraining/python/ort_trainer.py]: Yea, ORT saves both fp16 weights and fp32 weights. I got a bit confused earlier. But I decided to save a copy of original torch state keys and just grab the value from ort state using exact key match..

File: orttraining/orttraining/test/gradient/gradient_ops_test.cc (hunk 4 above)
Comment (Tixxx) [reply orttraining/orttraining/test/gradient/gradient_ops_test.cc:516]: refactored the code.

File: onnxruntime/core/providers/cuda/reduction/reduction_ops.cc (hunk 6 above)
Comment (Tixxx) [reply onnxruntime/core/providers/cuda/reduction/reduction_ops.cc]: updated

File: orttraining/orttraining/test/gradient/gradient_ops_test.cc [hunk 8]
```diff
@@ -401,149 +424,138 @@ TEST(GradientCheckerTest, GemmGrad) {
...
-  }
-
-  // axes = [-2, -1], keepdims = 0
-  {
-    gradient_checker.ComputeGradientError(op_def, {{4, 3, 2}}, {{4}}, &max_error,
-                                          {MakeAttribute("axes", std::vector<int64_t>{-2, -1}),
-                                           MakeAttribute("keepdims", int64_t(0))});
-    EXPECT_IS_TINY(max_error);
-  }
+  input_x_vector input_x;
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/gradient/gradient_ops_test.cc:430]: The test data can also be shared across the 3 ops, right? 
Move them into the RunReductionTests.

Comment (Tixxx) [reply orttraining/orttraining/test/gradient/gradient_ops_test.cc:430]: I exposed the testdate out in case other reduce ops in the future want to test with other data and attributes.

File: orttraining/orttraining/test/gradient/gradient_op_test_utils.h [hunk 9]
```diff
@@ -7,6 +7,9 @@
 
 namespace onnxruntime {
 namespace test {
+using input_x_vector = std::vector<std::vector<TensorInfo>>;
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/gradient/gradient_op_test_utils.h:10]: InputVector
OutputVector
AttrVector

See google C++ coding guideline:  https://google.github.io/styleguide/cppguide.html#Type_Names

File: orttraining/orttraining/python/ort_trainer.py [hunk 10]
```diff
@@ -773,7 +775,15 @@ def state_dict(self):
             if n.name not in torch_state:
                 torch_state[n.name] = torch.from_numpy(numpy_helper.to_array(n))
 
-        return torch_state
+        # Need to remove redundant initializers and name suffices to map back to original torch state names
+        torch_state_to_return = {}
```
Comment (SherlockNoMad) [inline orttraining/orttraining/python/ort_trainer.py:779]: I remember python have a syntax like this 

torch_state_to_return = [x if x in self.original_model_state_keys]

File: orttraining/orttraining/test/gradient/gradient_ops_test.cc (hunk 8 above)
Comment (SherlockNoMad) [reply orttraining/orttraining/test/gradient/gradient_ops_test.cc:430]: I don't think we will need those cases. Gradient checking for Reduce op will be very similar is not identical... 
If you really want to accommodate for the future cases, you can add "addition test data field" as the function inputs

File: orttraining/orttraining/python/ort_trainer.py (hunk 10 above)
Comment (Tixxx) [reply orttraining/orttraining/python/ort_trainer.py:779]: sure, i can change it to a dictionary comprehension.

File: orttraining/orttraining/test/gradient/gradient_op_test_utils.h [hunk 11]
```diff
@@ -7,6 +7,9 @@
 
 namespace onnxruntime {
 namespace test {
+using TestDataVector = std::tuple<std::vector<std::vector<TensorInfo>>, // Input data
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/gradient/gradient_op_test_utils.h:10]: I would recommend std::vector<std::tuple<x,y,attr>>, so each test case can be view together. 

But it's your call.
